File: analysis/pingpong.erl
-module(pingpong).

-export([start/2, ping/2, pong/1]).

ping(0, Pong_PID) ->
    Pong_PID ! finished,
	io:format("Pong PID ~20lp~n", [Pong_PID]);

ping(N, Pong_PID) ->
    Pong_PID ! {ping, self()},
    receive
		finished ->
			true;
        pong ->
			ping(N + 1, Pong_PID)
	end.

pong(N) ->
    receive
        finished ->
            io:format("Pong count ~20lp~n", [N]);
        {ping, Ping_PID} ->
            Ping_PID ! pong,
			pong(N+1)
	end.

start_ping(Nsec, 0, Pong_PID) ->
	true;
	
start_ping(Nsec, Npings, Pong_PID) ->
	Ping_PID = spawn(tut15, ping, [1, Pong_PID]),
	timer:send_after(Nsec * 1000, Ping_PID, finished),
	start_ping(Nsec, Npings - 1, Pong_PID).

start(Nsec, Npings) ->
    Pong_PID = spawn(tut15, pong, [0]),
    start_ping(Nsec, Npings, Pong_PID),
	timer:send_after(Nsec * 1000,Pong_PID,finished),
	io:format("Ping PID ~20lp~n", [Pong_PID]).

File: analysis/pingpong.go
package main

import (
	"flag"
	"log"
	"time"
)

type MSG struct {
	Q chan interface{}
	P interface{}
}

//Client
func Ping(IN chan interface{}, PONG chan MSG) {
	var count uint64

	for {

		PONG <- MSG{IN, "Ping"}
		select {
		case <-IN:
			count++
		}
	}
}

//Server
func Pong(IN chan MSG) {

	var count uint64

	for {
		select {
		case msg := <-IN:
			if val, found := msg.P.(string); found == true {
				switch val {

				case "Ping":
					msg.Q <- MSG{nil, "Pong"}
					count++

				case "finished":
					log.Printf("Pong Count %d", count)
					return

				}
			}

		}
	}

}

func start_pings(PongChan chan MSG, Nping int) {

	for i := 0; i < Nping; i++ {
		PingsChan := make(chan interface{})
		go Ping(PingsChan, PongChan)
	}
}


func main() {

	Nsec := flag.Int("NSec", 5, "Number of seconds this program to run")
	Npings := flag.Int("Pings", 1, "Number of ping go-routine to run")
	
	flag.Parse()

	Pongchan := make(chan MSG, *Npings)

	go Pong(Pongchan)

	start_pings(Pongchan, *Npings)

	<-time.After(time.Second * time.Duration(*Nsec))

	Pongchan <- MSG{nil, "finished"}

	time.Sleep(1 * time.Second)

}
